graphql_compiler/tests/interpreter_tests/test_interpreter_behavior.py

```diff
@@ -237,3 +240,27 @@ def test_single_generator_pull_grabs_only_one_result_from_adapter(self) -> None:
             )
         )
         self.assertEqual(expected_trace, trace)
+
+    def test_with_local_field(self) -> None:
+        adapter = InterpreterAdapterTap(InMemoryTestAdapter())
+
+        query = """{
+            Animal {
+                color @tag(tag_name: "color")
+                name @output(out_name: "name")
+                     @filter(op_name: "=", value: ["%color"])
+            }
+        }"""
+        args: Dict[str, Any] = {}
+
+        result_gen = interpret_query(adapter, self.schema, query, args)
+        list(result_gen)  # drain the iterator
+
+        trace = adapter.recorder.get_trace()
+
+        project_property_calls = [
+            operation
+            for operation in trace.operations
+            if operation.kind == "call" and operation.name == "project_property"
+        ]
+        self.assertEqual(3, len(project_property_calls))  # TODO(bojanserafimov): Why 3?
```

Comment:
@obi1kenobi There's 10 animals in the dataset, not sure how this completes with 3 calls to `project_property`

Reply:
Ah, you've stumbled upon precisely the part that makes the interpreter difficult to implement, but once implemented makes the API pleasant and powerful.

Look at the `field_name` values of the `project_property` calls. They are going to be `name`, `color`, and `name`, almost certainly in that order. Here's what's happening:
- First call (`name`): this is to get the value of the `name @filter(...)` part.
- Second call (`color`): this is to get a value for the `value: ["%color"]` tagged parameter inside the filter.
- Third call (`name`): this is to perform `name @output(...)`.

You correctly point out that there are 10 animals, and it would be a valid execution of the interpreter if it called `project_property` for each used property for each animal (30 calls in total). But it clearly doesn't do that.

To see why, recall the interpreter's desirable optimality ("anything you can do, I can do at least as well") property. This property doesn't hold if we can show an optimization that is prevented by the interpreter API's design, or by the interpreter library's implementation.

Making anything more than the 3 calls I described above does prevent a very desirable optimization: user-specified batching. Imagine `name` for animals 1-4 is fetched in one `project_property` call, and for animals 5-10 in another `project_property` call. This prevents the user from batching 10 (or 8, or all, or `random.randint()`) animals at a time — no implementation of the `InterpreterAdapter` API is able to batch over input data it is unable to receive as input, since the `data_contexts` input generator will run out.

So optimality requires that *all* the relevant `data_contexts` are delivered in a single call to `project_property` (or any of the other two functions that receive generators as input). That means that the set of function *calls* in the trace is purely a function of the input query, and does not depend on the data.

Taking a step back, *this* is why the interpreter is so challenging to build: because it has to dynamically construct a chain of generators in *just the right way* that fulfills the target query without knowing how the 4-method `InterpreterAdapter` API is going to be implemented, and by extension guaranteeing correct results while not knowing the exact order of operations between those generators beyond "data is created before it is consumed."